File: python/mirage/commands/cli/builtin/airtable/reads.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import functools

from mirage.accessor.airtable import AirtableAccessor
from mirage.commands.cli.builtin.airtable.util import (Outcome, no_operands,
                                                       one_operand, run,
                                                       scoped_base,
                                                       usage_error)
from mirage.commands.cli.types import CLIInvocation
from mirage.commands.spec.flag_view import FlagView
from mirage.core.airtable.client import (get_record, list_bases, list_comments,
                                         list_records, list_tables)
from mirage.core.airtable.config import AirtableConfig
from mirage.core.airtable.normalize import (normalize_base,
                                            normalize_base_summary,
                                            normalize_comment, normalize_table,
                                            records_jsonl, to_json_bytes)
from mirage.io.stream import yield_bytes
from mirage.io.types import IOResult


async def _base_list(accessor: AirtableAccessor,
                     inv: CLIInvocation[AirtableConfig], fl: FlagView,
                     prog: str) -> Outcome:
    no_operands(prog, inv.texts)
    bases = await list_bases(accessor)
    payload = [normalize_base_summary(base) for base in bases]
    return yield_bytes(to_json_bytes(payload)), IOResult()


async def _base_get(accessor: AirtableAccessor,
                    inv: CLIInvocation[AirtableConfig], fl: FlagView,
                    prog: str) -> Outcome:
    base_id = scoped_base(inv.config, one_operand(prog, inv.texts, "BASE"))
    for base in await list_bases(accessor):
        if base.get("id") == base_id:
            tables = await list_tables(accessor, base_id)
            data = to_json_bytes(normalize_base(base, tables))
            return yield_bytes(data), IOResult()
    raise LookupError(f"{base_id}: no such base")


async def _table_get(accessor: AirtableAccessor,
                     inv: CLIInvocation[AirtableConfig], fl: FlagView,
                     prog: str) -> Outcome:
    ref = one_operand(prog, inv.texts, "TABLE")
    base_id = scoped_base(inv.config, fl.as_str("base") or "")
    tables = await list_tables(accessor, base_id)
    table = (next((t for t in tables if t.get("id") == ref), None) or next(
        (t for t in tables if t.get("name") == ref), None))
    if table is None:
        raise LookupError(f"{ref}: no such table in {base_id}")
    return yield_bytes(to_json_bytes(normalize_table(table,
                                                     base_id))), IOResult()


async def _record_list(accessor: AirtableAccessor,
                       inv: CLIInvocation[AirtableConfig], fl: FlagView,
                       prog: str) -> Outcome:
    no_operands(prog, inv.texts)
    asked = fl.as_int("max_records")
    if asked is not None and asked < 1:
        raise usage_error(prog, "--max-records must be at least 1")
    base_id = scoped_base(inv.config, fl.as_str("base") or "")
    cap = inv.config.max_read_records
    records = await list_records(accessor,
                                 base_id,
                                 fl.as_str("table") or "",
                                 view=fl.as_str("view"),
                                 formula=fl.as_str("formula"),
                                 max_records=cap +
                                 1 if asked is None else asked)
    if asked is None and len(records) > cap:
        raise ValueError(f"more than {cap} records match (max_read_records);"
                         " narrow them with --formula or --view, or take the"
                         " first N with --max-records N")
    return yield_bytes(records_jsonl(records)), IOResult()


async def _record_get(accessor: AirtableAccessor,
                      inv: CLIInvocation[AirtableConfig], fl: FlagView,
                      prog: str) -> Outcome:
    record_id = one_operand(prog, inv.texts, "RECORD")
    base_id = scoped_base(inv.config, fl.as_str("base") or "")
    record = await get_record(accessor, base_id,
                              fl.as_str("table") or "", record_id)
    return yield_bytes(records_jsonl([record])), IOResult()


async def _comment_list(accessor: AirtableAccessor,
                        inv: CLIInvocation[AirtableConfig], fl: FlagView,
                        prog: str) -> Outcome:
    record_id = one_operand(prog, inv.texts, "RECORD")
    base_id = scoped_base(inv.config, fl.as_str("base") or "")
    comments = await list_comments(accessor, base_id,
                                   fl.as_str("table") or "", record_id)
    payload = [normalize_comment(comment) for comment in comments]
    return yield_bytes(to_json_bytes(payload)), IOResult()


base_list = functools.partial(run, "base list", _base_list)
base_get = functools.partial(run, "base get", _base_get)
table_get = functools.partial(run, "table get", _table_get)
record_list = functools.partial(run, "record list", _record_list)
record_get = functools.partial(run, "record get", _record_get)
comment_list = functools.partial(run, "comment list", _comment_list)

File: typescript/packages/core/src/commands/cli/builtin/airtable/reads.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { AirtableAccessor } from '../../../../accessor/airtable.ts'
import {
  getRecord,
  listBases,
  listComments,
  listRecords,
  listTables,
} from '../../../../core/airtable/client.ts'
import type { AirtableConfig } from '../../../../core/airtable/config.ts'
import {
  normalizeBase,
  normalizeBaseSummary,
  normalizeComment,
  normalizeTable,
  recordsJsonl,
  toJsonBytes,
} from '../../../../core/airtable/normalize.ts'
import { IOResult } from '../../../../io/types.ts'
import type { CommandFnResult } from '../../../config.ts'
import type { FlagView } from '../../../spec/flag_view.ts'
import type { CLIInvocation } from '../../types.ts'
import { PROG, noOperands, oneOperand, run, scopedBase, usageError } from './util.ts'

function config(inv: CLIInvocation): AirtableConfig {
  return inv.config as AirtableConfig
}

async function baseListBody(
  accessor: AirtableAccessor,
  inv: CLIInvocation,
  _fl: FlagView,
): Promise<CommandFnResult> {
  noOperands(`${PROG} base list`, inv.texts)
  const bases = await listBases(accessor)
  return [toJsonBytes(bases.map(normalizeBaseSummary)), new IOResult()]
}

async function baseGetBody(
  accessor: AirtableAccessor,
  inv: CLIInvocation,
  _fl: FlagView,
): Promise<CommandFnResult> {
  const baseId = scopedBase(config(inv), oneOperand(`${PROG} base get`, inv.texts, 'BASE'))
  for (const base of await listBases(accessor)) {
    if (base.id === baseId) {
      const tables = await listTables(accessor, baseId)
      return [toJsonBytes(normalizeBase(base, tables)), new IOResult()]
    }
  }
  throw new Error(`${baseId}: no such base`)
}

async function tableGetBody(
  accessor: AirtableAccessor,
  inv: CLIInvocation,
  fl: FlagView,
): Promise<CommandFnResult> {
  const ref = oneOperand(`${PROG} table get`, inv.texts, 'TABLE')
  const baseId = scopedBase(config(inv), fl.asStr('base') ?? '')
  const tables = await listTables(accessor, baseId)
  const table = tables.find((t) => t.id === ref) ?? tables.find((t) => t.name === ref)
  if (table === undefined) throw new Error(`${ref}: no such table in ${baseId}`)
  return [toJsonBytes(normalizeTable(table, baseId)), new IOResult()]
}

async function recordListBody(
  accessor: AirtableAccessor,
  inv: CLIInvocation,
  fl: FlagView,
): Promise<CommandFnResult> {
  const prog = `${PROG} record list`
  noOperands(prog, inv.texts)
  const asked = fl.asInt('max_records')
  if (asked !== undefined && asked < 1) throw usageError(prog, '--max-records must be at least 1')
  const baseId = scopedBase(config(inv), fl.asStr('base') ?? '')
  const cap = accessor.maxReadRecords
  const view = fl.asStr('view')
  const formula = fl.asStr('formula')
  const records = await listRecords(accessor, baseId, fl.asStr('table') ?? '', {
    ...(view !== undefined ? { view } : {}),
    ...(formula !== undefined ? { formula } : {}),
    maxRecords: asked ?? cap + 1,
  })
  if (asked === undefined && records.length > cap) {
    throw new Error(
      `more than ${String(cap)} records match (max_read_records); narrow them with --formula ` +
        'or --view, or take the first N with --max-records N',
    )
  }
  return [recordsJsonl(records), new IOResult()]
}

async function recordGetBody(
  accessor: AirtableAccessor,
  inv: CLIInvocation,
  fl: FlagView,
): Promise<CommandFnResult> {
  const recordId = oneOperand(`${PROG} record get`, inv.texts, 'RECORD')
  const baseId = scopedBase(config(inv), fl.asStr('base') ?? '')
  const record = await getRecord(accessor, baseId, fl.asStr('table') ?? '', recordId)
  return [recordsJsonl([record]), new IOResult()]
}

async function commentListBody(
  accessor: AirtableAccessor,
  inv: CLIInvocation,
  fl: FlagView,
): Promise<CommandFnResult> {
  const recordId = oneOperand(`${PROG} comment list`, inv.texts, 'RECORD')
  const baseId = scopedBase(config(inv), fl.asStr('base') ?? '')
  const comments = await listComments(accessor, baseId, fl.asStr('table') ?? '', recordId)
  return [toJsonBytes(comments.map(normalizeComment)), new IOResult()]
}

export const baseList = run(baseListBody)
export const baseGet = run(baseGetBody)
export const tableGet = run(tableGetBody)
export const recordList = run(recordListBody)
export const recordGet = run(recordGetBody)
export const commentList = run(commentListBody)
